conversation/deepseek.go

```diff
@@ -0,0 +1,201 @@
+}
+
+type DeepSeekLogprobs struct {
+	Content []DeepSeekLogprobContent `json:"content"`
+}
+
+type DeepSeekLogprobContent struct {
+	Token       string                   `json:"token"`
+	Logprob     float64                  `json:"logprob"`
+	Bytes       []byte                   `json:"bytes"`
+	TopLogprobs []DeepSeekTopLogprobItem `json:"top_logprobs"`
+}
+
+type DeepSeekTopLogprobItem struct {
+	Token   string  `json:"token"`
+	Logprob float64 `json:"logprob"`
+	Bytes   []byte  `json:"bytes"`
+}
+
+type DeepSeekUsage struct {
+	CompletionTokens      int `json:"completion_tokens"`
+	PromptTokens          int `json:"prompt_tokens"`
+	PromptCacheHitTokens  int `json:"prompt_cache_hit_tokens"`
+	PromptCacheMissTokens int `json:"prompt_cache_miss_tokens"`
+	TotalTokens           int `json:"total_tokens"`
+}
+
+func (p *DeepSeekParticipant) GenerateResponse(ctx context.Context, history []Message) (string, error) {
+	type messagePayload struct {
+		Role    string `json:"role"`
+		Content string `json:"content"`
+	}
+	var messagesPayload []messagePayload
+	for i := range history {
+		var role string
+		if strings.Compare(strings.ToLower(history[i].From), "system") == 0 {
+			role = "system"
+		} else if strings.Compare(strings.ToLower(history[i].From), strings.ToLower(p.GetName())) == 0 {
+			role = "assistant"
+		} else {
+			role = "user"
+		}
+		messagesPayload = append(messagesPayload, messagePayload{Role: role, Content: history[i].Content})
+	}
+	payloadMap := map[string]interface{}{
+		"messages":          messagesPayload,
+		"model":             "deepseek-chat",
+		"frequency_penalty": 0,
+		"max_tokens":        2048,
+		"presence_penalty":  0,
+		"response_format": map[string]interface{}{
+			"type": "text",
+		},
+		"stop":           nil,
+		"stream":         false,
+		"stream_options": nil,
+		"temperature":    1,
+		"top_p":          1,
+		"tools":          nil,
+		"tool_choice":    "none",
+		"logprobs":       false,
+		"top_logprobs":   nil,
+	}
+	payload, err := json.Marshal(payloadMap)
+	if err != nil {
+		return "", err
+	}
+	client := &http.Client{}
+	req, err := http.NewRequest("POST", "https://api.deepseek.com/chat/completions", bytes.NewReader(payload))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", p.apiKey))
+
+	res, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", err
+	}
+	var chatCompletion DeepSeekChatCompletion
+	err = json.Unmarshal([]byte(body), &chatCompletion)
+	if err != nil {
+		return "", fmt.Errorf("error while unrmasaling json: %w, body: %s", err, string(body))
+	}
+
+	return chatCompletion.Choices[0].Message.Content, nil
+}
+
+func (p *DeepSeekParticipant) manageConversation(messages []openai.ChatCompletionMessage) []openai.ChatCompletionMessage {
```

Comment:
Looks like you missed checking that we are not exceeding the context window. `manageConversation` is not getting called anywhere.
` manageConversation(messages []openai.ChatCompletionMessage) []openai.ChatCompletionMessage  ` needs to be updated to deepseek struct.

Reply:
Please do update readme to include deepseek as well, thanks.

Reply:
Is this good implementation?
```golang
history = append(history, Message{
From:    p.GetName(),
Content: chatCompletion.Choices[0].Message.Content,
})

history = p.manageConversation(history)

return history[len(history)-1].Content, nil
```

Reply:
Similar to other implementations:

```
var messagesPayload []messagePayload
	for i := range history {
		var role string
		if strings.Compare(strings.ToLower(history[i].From), "system") == 0 {
			role = "system"
		} else if strings.Compare(strings.ToLower(history[i].From), strings.ToLower(p.GetName())) == 0 {
			role = "assistant"
		} else {
			role = "user"
		}
		messagesPayload = append(messagesPayload, messagePayload{Role: role, Content: history[i].Content})
	}
	
// handle context window
messagesPayload = p.manageConversation(messagesPayload)
		

```

Update manageConversation accordingly.